File: graph_querier.py
"""
Knowledge Graph Querier Module
Query the knowledge graph using SPARQL
"""

from typing import List, Dict
from rdflib import Literal, RDFS
from graph_builder import KnowledgeGraphBuilder


class KnowledgeGraphQuerier:
    """Query the knowledge graph using SPARQL"""
    
    def __init__(self, kg_builder: KnowledgeGraphBuilder):
        self.kg = kg_builder
    
    def query(self, sparql_query: str) -> List[Dict]:
        """Execute SPARQL query"""
        results = []
        qres = self.kg.graph.query(sparql_query)
        
        for row in qres:
            result_dict = {}
            for var in qres.vars:
                value = row[var]
                if isinstance(value, Literal):
                    result_dict[str(var)] = str(value)
                else:
                    label = self.kg.graph.value(value, RDFS.label)
                    result_dict[str(var)] = str(label) if label else str(value)
            results.append(result_dict)
        
        return results
    
    def find_entity_relations(self, entity_name: str) -> List[Dict]:
        """Find all relations for an entity"""
        query = f"""
        PREFIX kg: <{self.kg.ns}>
        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
        
        SELECT ?predicate ?object ?objLabel
        WHERE {{
            ?subject rdfs:label "{entity_name}" .
            ?subject ?predicate ?object .
            OPTIONAL {{ ?object rdfs:label ?objLabel }}
        }}
        """
        return self.query(query)

File: backend/src/main/kotlin/io/github/juantoanxoup/kg/GraphQuerier.kt
package io.github.juantoanxoup.kg

import org.apache.jena.query.ParameterizedSparqlString
import org.apache.jena.query.QueryExecutionFactory
import org.apache.jena.query.QueryFactory
import org.apache.jena.rdf.model.RDFNode
import org.apache.jena.vocabulary.RDFS

/** SPARQL access to the knowledge graph (graph_querier.py). */
class KnowledgeGraphQuerier(
    private val kg: KnowledgeGraphBuilder,
) {
    /**
     * Executes a SELECT query. Literals become their lexical form; IRIs become their `rdfs:label`
     * when one exists, else the IRI string. Unbound variables become an empty string.
     */
    fun query(sparql: String): List<Map<String, String>> {
        val parsed = QueryFactory.create(sparql)
        require(parsed.isSelectType) { "Only SELECT queries are supported" }
        return QueryExecutionFactory.create(parsed, kg.model).use { exec ->
            val results = exec.execSelect()
            val vars = results.resultVars
            results.asSequence().map { row -> vars.associateWith { render(row[it]) } }.toList()
        }
    }

    /** All outgoing relations of the entity whose `rdfs:label` equals [entityName]. */
    fun findEntityRelations(entityName: String): List<Map<String, String>> {
        val query =
            ParameterizedSparqlString(
                """
                PREFIX kg: <${kg.namespace}>
                PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

                SELECT ?predicate ?object ?objLabel
                WHERE {
                    ?subject rdfs:label ?name .
                    ?subject ?predicate ?object .
                    OPTIONAL { ?object rdfs:label ?objLabel }
                }
                """.trimIndent(),
            )
        query.setLiteral("name", entityName)
        return query(query.toString())
    }

    private fun render(node: RDFNode?): String =
        when {
            node == null -> ""
            node.isLiteral -> node.asLiteral().lexicalForm
            else ->
                kg.model
                    .getProperty(node.asResource(), RDFS.label)
                    ?.`object`
                    ?.asLiteral()
                    ?.lexicalForm
                    ?: node.toString()
        }
}
